src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -69,9 +71,21 @@ def resolve(self, root_reqs, check_supported_wheels):
         # The persistent state that we care about is `root_reqs`.
         assert len(self.factory.root_reqs) == 0, "Factory is being re-used"
 
-        # FIXME: Implement constraints.
-        if any(r.constraint for r in root_reqs):
-            raise InstallationError("Constraints are not yet supported.")
+        constraints = defaultdict(list)  # type: Dict[str,List[SpecifierSet]]
```

Comment:
This uses `defaultdict` but the provider uses `dict.get()` to access. Only one of them is required.

Reply:
Never though of that! I typed it as `Dict` though - does that matter? I'm using `defaultdict` because it's easier for construction, but once constructed I want to treat it as a dictionary (although I'm fine with omitting the `get` call).

Reply:
Hmm, yeah I guess that makes sense if you type it as `Dict`. I *think* `defaultdict.__getitem__()` is more performance than `dict.get()`, but that’s too insignificant to matter anyway. Let’s keep this as-is.

Reply:
Hmm. On reflection, I'd rather keep the `.get()`, just in case someone later refactors the code that builds the constraints and uses a normal dictionary rather than a defaultdict.